constraint-solver/src/solver.rs

```diff
@@ -96,8 +110,187 @@ pub enum Error {
+    }
+}
+
+impl<T, V, S> Solver<T, V> for BooleanExtractedSolver<T, V, S>
+where
+    T: RuntimeConstant + VarTransformable<V, Variable<V>> + Display,
+    T::Transformed: RuntimeConstant<FieldType = T::FieldType>
+        + VarTransformable<Variable<V>, V, Transformed = T>
+        + Display,
+    V: Ord + Clone + Eq + Hash + Display,
+    S: Solver<T::Transformed, Variable<V>>,
+{
+    /// Solves the system and ignores all assignments that contain a boolean variable
+    /// (either on the LHS or the RHS).
+    fn solve(&mut self) -> Result<Vec<VariableAssignment<T, V>>, Error> {
+        let assignments = self.solver.solve()?;
+        Ok(assignments
+            .into_iter()
+            .filter_map(|(v, expr)| {
+                let v = v.try_to_original()?;
+                let expr = expr.try_transform_var_type(&mut |v| v.try_to_original())?;
+                Some((v, expr))
+            })
+            .collect())
+    }
+
+    fn add_algebraic_constraints(
+        &mut self,
+        constraints: impl IntoIterator<Item = GroupedExpression<T, V>>,
+    ) {
+        let constraints = constraints
+            .into_iter()
+            .flat_map(|constr| {
+                let constr = constr.transform_var_type(&mut |v| v.clone().into());
+                let extracted =
+                    try_extract_boolean(&constr, &mut || self.boolean_var_dispenser.next_var())
+                        .inspect(|_| {
+                            // Make sure to range-constrain the variable
+                            self.solver.add_range_constraint(
+                                &self.boolean_var_dispenser.latest_var().unwrap(),
+                                RangeConstraint::from_mask(1),
+                            );
+                        });
+                std::iter::once(constr).chain(extracted)
+            })
+            .collect_vec();
+        self.solver.add_algebraic_constraints(constraints);
+    }
+
+    fn add_bus_interactions(
+        &mut self,
+        bus_interactions: impl IntoIterator<Item = BusInteraction<GroupedExpression<T, V>>>,
+    ) {
+        self.solver
+            .add_bus_interactions(bus_interactions.into_iter().map(|bus_interaction| {
+                bus_interaction
+                    .fields()
+                    .map(|expr| {
+                        // We cannot extract booleans here because that only works
+                        // for "constr = 0".
+                        expr.transform_var_type(&mut |v| v.clone().into())
+                    })
+                    .collect()
+            }))
+    }
+
+    fn add_range_constraint(&mut self, variable: &V, constraint: RangeConstraint<T::FieldType>) {
+        self.solver
+            .add_range_constraint(&variable.clone().into(), constraint);
+    }
+
+    fn retain_variables(&mut self, variables_to_keep: &HashSet<V>) {
+        let variables_to_keep = variables_to_keep
+            .iter()
+            .map(|v| Variable::from(v.clone()))
```

Comment:
Wait, but this removes all boolean variables. Is this what we want?

Reply:
I actually misunderstood, I though `retain_variables` removes a constraint unless it only references a variable in `variables_to_keep`.

Do I understand this right that the `retain_variables` is purely performance optimization? Doing nothing would yield the same result?

Reply:
The old description was misleading yes. I hope it's better now. And yes, it's only about performance. You should not remove variables that are connected to stateful bus interactions.